presto-hive/src/main/java/io/prestosql/plugin/hive/ViewReaderUtil.java

```diff
@@ -0,0 +1,229 @@
+import io.prestosql.spi.PrestoException;
+import io.prestosql.spi.connector.ConnectorViewDefinition;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.TypeManager;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.hadoop.hive.metastore.TableType;
+
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+
+import static io.prestosql.plugin.hive.HiveErrorCode.HIVE_INVALID_VIEW_DATA;
+import static io.prestosql.plugin.hive.HiveErrorCode.HIVE_UNKNOWN_ERROR;
+import static io.prestosql.plugin.hive.HiveMetadata.TABLE_COMMENT;
+import static io.prestosql.plugin.hive.util.HiveUtil.checkCondition;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static org.apache.hadoop.hive.metastore.TableType.VIRTUAL_VIEW;
+
+public final class ViewReaderUtil
+{
+    private ViewReaderUtil()
+    {}
+
+    public interface ViewReader
+    {
+        ConnectorViewDefinition decodeViewData(String viewData, Table table, CatalogName catalogName);
+    }
+
+    public static ViewReader createViewReader(SemiTransactionalHiveMetastore metastore, HiveIdentity identity, Table table, TypeManager typemanager)
+    {
+        if (isPrestoView(table)) {
+            return new io.prestosql.plugin.hive.ViewReaderUtil.PrestoViewDecoder();
+        }
+        else {
+            return new io.prestosql.plugin.hive.ViewReaderUtil.HiveViewDecoder(new CoralSemiTransactionalHiveMSCAdapter(metastore, identity), typemanager);
+        }
+    }
+
+    public static final String PRESTO_VIEW_FLAG = "presto_view";
+    static final String VIEW_PREFIX = "/* Presto View: ";
+    static final String VIEW_SUFFIX = " */";
+    private static final JsonCodec<ConnectorViewDefinition> VIEW_CODEC =
+            new JsonCodecFactory(new ObjectMapperProvider()).jsonCodec(ConnectorViewDefinition.class);
+    private static Logger log = Logger.get(io.prestosql.plugin.hive.ViewReaderUtil.class);
+
+    public static boolean isPrestoView(Table table)
+    {
+        return "true".equals(table.getParameters().get(PRESTO_VIEW_FLAG));
+    }
+
+    public static boolean isHiveOrPrestoView(Table table)
+    {
+        return table.getTableType().equals(TableType.VIRTUAL_VIEW.name());
+    }
+
+    public static boolean canDecodeView(Table table)
+    {
+        // we can decode Hive or Presto view
+        return table.getTableType().equals(VIRTUAL_VIEW.name());
+    }
+
+    public static String encodeViewData(ConnectorViewDefinition definition)
+    {
+        byte[] bytes = VIEW_CODEC.toJsonBytes(definition);
+        String data = Base64.getEncoder().encodeToString(bytes);
+        return VIEW_PREFIX + data + VIEW_SUFFIX;
+    }
+
+    /**
+     * Supports decoding of Presto views
+     */
+    public static class PrestoViewDecoder
```

Comment:
Let's call these `PrestoViewReader` and `HiveViewReader` so that the names match the interface (that's the typical convention).